### PR TITLE
Resize outputs possibly considering strides.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67016
* __->__ #67015

**The Problem:** `resize_output` should be (re)allocating memory taking into account the combination of sizes and strides (see https://github.com/pytorch/pytorch/issues/65863#issuecomment-930784119)

**What was wrong with `resize_impl_{cpu,cuda}_`?** those functions reallocated memory even if the sizes (new and old) were the same (i.e. contiguous tensor fits the already allocated memory).

**The Solution:** in order not to break old code that relied on this behavior, I implemented a new way of computing the needed bytes for the new storage (which ultimately dictates whether the tensor will be reallocated or not), ignoring the given `strides` if necessary.

-----

In summary, this PR:
- Introduces an overload of `resize_` kernel that accepts `strides` as parameter
- Unifies the implementation of `resize_` and `resize_cuda_` with `resize_template`
- Unifies the implementation of `resize_impl_{cpu,cuda}_` with `resize_impl_template_`
- Modifies `resize_output` for working with optional `strides`
- Make codegen for structured kernels pass `strides` when resizing a given out tensor

-----

`resize_impl_template_` summarizes both `resize_impl_{cpu,cuda}_` in one function. In order to use it, one must provide 2 template parameters:
- The function that may actually resize the tensor (`MaybeResize` -- different for CPU and CUDA)
- The function that will compute the new size in bytes (`ComputeStorageSize`)

Doing so, we can stick to the same `resize_impl_template_` function while using a different strategy for computing the storage size needed. That was needed in `resize_output`, when porting operations to be a structured kernel . The new strategy `compute_storage_size_tryreuse` checks if the new tensor would fit the existing storage with 2 strategies, before actually deciding whether to re-allocate or not:

1. using the requested strides
2. ignoring the requested strides (making it contiguous)

If none of these 2 strategies work (allow reuse), the function fallbacks to re-allocating new memory, giving priority to strategy 1 (use requested strides).

### CODE REVIEW

**Comment by ysiraichi** [comment]
@ezyang @bdhirsh @krshrimali 
Could you guys take a look at this PR? The 2 CI failures aren't related to this PR.
I'm not completely sure we want this behavior for `resize_outputs`, but this was what made the most sense to me. Let me know what you guys think.

**Comment by bdhirsh** [comment]
cc @ngimel for your opinion on the `resize_` changes. I'm also a little less familiar with the existing `resize_` logic.

File: aten/src/ATen/native/Resize.cpp [hunk 1]
```diff
@@ -35,18 +35,21 @@ static bool is_functorch_wrapped_tensor(const Tensor& tensor) {
   return !(key_set & kFunctorchWrappedTensors).empty();
 }
 
-bool resize_output(const Tensor& output, IntArrayRef shape) {
+bool resize_output(const Tensor& output, IntArrayRef shape, c10::optional<IntArrayRef> strides) {
```
Comment (bdhirsh) [inline aten/src/ATen/native/Resize.cpp:38]: `resize_output` is also used in a bunch of places outside of codegen, so maybe we should default strides to `c10::nullopt`?

File: aten/src/ATen/native/Resize.cpp [hunk 2]
```diff
@@ -35,18 +35,21 @@ static bool is_functorch_wrapped_tensor(const Tensor& tensor) {
...
   if (resize_output_check(output, shape)) {
-    // avoid a redispatch for cpu and cuda.
-    // TODO: when resize_cuda_ is re-written to be unified with resize_,
-    // we can provide the same benefit for cuda.
-    //
     // TODO(#61485): functorch wrapped tensors should not go through the
     // fast path. This is a hack, longer term solutions are in the issue
     if (output.is_cpu() && !is_functorch_wrapped_tensor(output)) {
-      at::native::resize_(output, shape);
+      at::native::resize_template<&resize_impl_tryreuse_<&maybe_resize_storage_cpu>>(
```
Comment (bdhirsh) [inline aten/src/ATen/native/Resize.cpp:43]: Is there any reason not to do something like
```
      if (strides.has_value()) {
        at::native::resize_(output, shape, strides.value());
      } else {
        at::native::resize_(output, shape);
      }
```

It's a little daunting to stare at that template when trying to understand the fastpath logic (but all we really care about is skipping the dispatch hop)

**Comment by ngimel** [comment]
I don't understand the description of the problem - old implementation didn't reallocate memory if existing storage was already enough:
```
In [2]: a=torch.randn(5,5)

In [3]: a.data_ptr()
Out[3]: 94259367559744

In [4]: a.resize_(25)
Out[4]: 
tensor([ 0.7261, -0.4577,  0.3521,  0.1484, -0.4512, -0.0693,  0.4181,  1.6139,
        -1.4124, -0.4129,  0.2849, -2.6132, -0.3775, -0.9499,  0.0239, -0.0866,
        -1.2876, -0.2752,  0.3847, -0.0800, -0.7892, -1.0268, -2.1730,  0.2168,
         0.5534])

In [5]: a.size()
Out[5]: torch.Size([25])

In [6]: a.data_ptr()
Out[6]: 94259367559744
```

File: aten/src/ATen/native/Resize.cpp [hunk 3]
```diff
@@ -122,25 +125,30 @@ const Tensor& resize_as_(
...
-  if (optional_memory_format.has_value()) {
-    auto memory_format =
-        optional_memory_format.value();
-    TORCH_CHECK(
-        memory_format != MemoryFormat::Preserve,
-        "Unsupported memory format",
-        memory_format);
-    self_->empty_tensor_restride(memory_format);
-  }
+  resize_template<&resize_impl_cpu_>(self, size, c10::nullopt, optional_memory_format, true);
```
Comment (krshrimali) [inline aten/src/ATen/native/Resize.cpp:142]: nit: it will be nice to pass the keywords of these arguments like: `/*arg_name=*/ true`, for code readability.

```cpp
resize_template<&resize_impl_cpu_>(self, size, /*strides=*/ c10::nullopt, optional_memory_format, /*resize_storage=*/ true);
```

**Review by krshrimali — COMMENTED** [review]
Hi, @ysiraichi - I, unfortunately, don't have much idea about the resize code, but I left a minor comment for readability. Feel free to ignore or address it later once others have reviewed the PR.

One small question, did you get a chance to benchmark the `resize_` function from this branch? Maybe just making sure that there is no perf regression, is a good idea, since it's been used a _lot_ of times. :)

**Comment by ngimel** [comment]
I second @krshrimali's suggestion, you can benchmark resize_ performance using instruction counts for deterministic results.

File: aten/src/ATen/native/Resize.cpp (hunk 1 above)
Comment (ysiraichi) [reply aten/src/ATen/native/Resize.cpp:38]: I also think so. The header file already does that!

File: aten/src/ATen/native/Resize.cpp (hunk 2 above)
Comment (ysiraichi) [reply aten/src/ATen/native/Resize.cpp:43]: I thought that by using the templates directly, I wouldn't need the branch. But, now that you mentioned it, I think it's way better to leave a branch there.

**Comment by ysiraichi** [comment]
@ngimel 
> I don't understand the description of the problem - old implementation didn't reallocate memory if existing storage was already enough:

In fact, there are 2 related problems, here:
1. There's no `resize_` that takes optional strides as arguments
2. `resize_` implementation (i.e. the functions `resize_impl_{cpu,cuda}_`) do take optional strides, but its behavior is different from what we want (see below)

-----
#### (1) `resize_` should take optional strides

Because of #65863, we wanted `resize_output` to take optional strides. That implied in having a `resize_` operation that also took optional strides (AFAIU, functorch needs it to go through the dispatcher).

----
#### (2) Different behavior of `resize_impl_{cpu,cuda}_`

Passing strides to `resize_impl_{cpu,cuda}_` had the following behavior:

```python
>>> a = torch.randn(5, 5)
>>> a.data_ptr()
<addr-1>
>>> a.resize_([25], strides=[10]) # using 'resize_impl_cpu_` internally
>>> a.data_ptr() # reallocation happens because strides differ
<addr-2>
```

The "culprit" being [this line](https://github.com/pytorch/pytorch/blob/e32d7f7525fded2c044dd690d3a1f52aa69ae79e/aten/src/ATen/native/Resize.h#L67-L69). Since some implementations may be relying on it, I didn't simply removed the strides from the condition. Instead, I implemented `resize_impl_tryreuse_` that gives:

```python
>>> a = torch.randn(5, 5)
>>> a.data_ptr()
<addr-1>
>>> a.resize_([25], strides=[10]) # using the new 'resize_impl_tryreuse_` internally
>>> a.data_ptr() # no allocation happens! It ignores the strides, since it won't fit otherwise
<addr-1>
```

**Comment by ngimel** [comment]
Sorry, I still don't understand it. If you start with (5,5) tensor you allocate storage enough for 25 elements. If you want to later resize it to ([25], strides=[10]), you need storage enough for 241 elements, so there's no way you could keep using your old storage.

**Comment by ysiraichi** [comment]
@ngimel Maybe the semantics I laid out is a bit confusing.

> If you want to later resize it to ([25], strides=[10]), you need storage enough for 241 elements

That is true if by `t.resize_([25], strides=[10])` we meant: _resize `t` to be a 1-D tensor with 25 elements and a stride of `[10]`_.

The reason why I implemented the new behavior (semantics) was because I think we wanted ([since that's how it was implemented](https://github.com/pytorch/pytorch/blob/04fe4382ec8e6031d353530d6b8446ac9c9f507e/tools/codegen/dest/register_dispatch_key.py#L76-L87)) in the context of structured kernels. In other words, the idea was to interpret `strides` as a suggestion. If we can use it without reallocating, great! Otherwise, just let it be contiguous.

The exact semantics of `t.resize_([25], strides=[10])` is: _resize `t` to be a 1-D tensor with 25 elements. Try making its stride `[10]`. If that doesn't fit the already allocated memory, make it contiguous. If that also doesn't fit and we need to reallocate anyway, make the new tensor with the given stride of `[10]`_.

-----
This PR does **NOT**:
- changes current `resize_(sizes, memory_format)` behavior
- changes current `resize_impl_{cpu,cuda}_` behavior

This PR does:
- add a new resize behavior (the one I explained above)
- exposes this new resize behavior with the operation `resize_(sizes, strides)` (maybe I should change the name to avoid confusion)

**Comment by ysiraichi** [comment]
@ngimel 
After some thought, I realized I wasn't really talking about the problem, but the solution I implemented. So, let me talk a little bit more about the problem, so that we can get to a possibly better solution.

----

### The Problem

It mainly comes with #65863. In a nutshell: calling [`set_output`](https://github.com/pytorch/pytorch/blob/94b6fa6f8b0cd730ad7401dc52880a9eb8a6e405/aten/src/ATen/TensorMeta.h#L60) with `strides` for the out-of-place overload of a structured kernel (e.g. `sort`) raises errors. Here's an example:

```python
>>> x = torch.randn(3, 5, dtype=torch.float)
>>> b = torch.empty(1)
>>> c = torch.empty(0, dtype=torch.long)
>>> torch.sort(x[:, 0], out=(b, c))
RuntimeError: setStorage: sizes [3], strides [5], storage offset 0, and itemsize 4 requiring a storage size of 44 are out of bounds for storage of size 12
```

That happened because, in the meta function for `sort`, we called `set_output` like:

```c++
// If the outputs have to be resized, do it while using 'self' strides.
set_output(0, self.sizes(), self.strides(), self.options(), {});
set_output(1, self.sizes(), self.strides(), self.options().dtype(kLong), {});
```

`set_outputs` is implemented by calling [`resize_out`](https://github.com/pytorch/pytorch/blob/94b6fa6f8b0cd730ad7401dc52880a9eb8a6e405/tools/codegen/dest/register_dispatch_key.py#L71). And **here** is where we have the real problem: 

```c++
// Possibly allocate new contiguous memory of size 'sizes'
const bool resized = at::native::resize_output(out, sizes);
if (resized) {
  if (!strides.empty()) {
    // 'as_strided_' errors out if the 'sizes' x 'strides' combination doesn't fit the allocated memory
    at::native::as_strided_(out, sizes, strides);
  } else if ...
}
```

For which @bdhirsh [suggested](https://github.com/pytorch/pytorch/issues/65863#issuecomment-930784119) to have `resize_output` take strides information together. Then, we would be able to use the strides if the output tensor were going to get resized anyway.

-----

### The Solutions

Here I will try to list the 2 solutions that make sense to me:

1. Tweak the implementation of `resize_` (i.e. `resize_impl_{cpu,cuda}_`), so that we use the strides conditionally (this PR)
2. Make `resize_output` use the strides conditionally

The reason why I went for (1) instead of (2) is that:

- (1) gives us finer control on when to use the strides
- (2) can't catch for errors (casting overflow errors)

Let me know what you guys think! Should we stick to (2)? Is there something else we can do?

**Comment by bdhirsh** [comment]
@ysiraichi I think @ngimel is calling out that just because structured kernels have specific semantics around resizing, that doesn't mean that we should change the semantics of `at::Tensor::resize_()`. The current semantics of `at::Tensor::resize_()` are that when you pass it strides, those strides are **not** a suggestion.

@ngimel I think the crux of the problem that we're trying to solve comes from these two different cases:
```
# Case 1: restriding happens
input = torch.empty((2, 2, 2, 2), memory_format=torch.channels_last)
output = torch.ones(0)
torch._C._nn.upsample_nearest2d(input, (2,2), out=output)
output.size() # torch.Size([2, 2, 2, 2])
# output was resized, AND restrided to be the same as input strides
output.stride() # (8, 1, 4, 2)

# Case 2: no restriding happens
x = torch.randn(3, 5, dtype=torch.float)
b = torch.empty(0)
c = torch.empty(0, dtype=torch.long)
torch.sort(x[:, 0], out=(b, c))
b.size() # torch.Size([3])
# output was resized, but is NOT restrided - it's contiguous.
b.stride() # (1,)
```

I'm not sure if that behavior is fully documented, but the main difference is that in case (1) it's easy to do the restriding since the underlying storage size is the same as the contiguous case, but in case (2) it's hard to do the restriding since that would require allocating a different amount of storage than the contiguous case. Does that seem like the right set of behavior?

If we're agreed on that, then @ysiraichi maybe we should just add a check in `resize_out` that first makes sure that restriding wouldn't require a storage size change before calling `as_strided_()`.

**Comment by ysiraichi** [comment]
Thanks for weighing in, @bdhirsh.

> The current semantics of `at::Tensor::resize_()` are that when you pass it strides, those strides are not a suggestion.

Just one thing here: `at::Tensor::resize_` doesn't take strides (up to this PR). I created it because I thought we needed this dispatch for _functorch_ and CUDA when calling `resize_output`.

> but in case (2) it's hard to do the restriding since that would require allocating a different amount of storage than the contiguous case

That's right! This is the reason why I had to change `resize_impl_{cpu,cuda}_` in this PR.

> maybe we should just add a check in `resize_out` that first makes sure that restriding wouldn't require a storage size change before calling `as_strided_()`?

That would be a solution, but is that the behavior we want? I mean, consider the following example:

```python
x = torch.randn(3, 5, dtype=torch.float)
b = torch.empty(0)
c = torch.empty(0, dtype=torch.long)
torch.sort(x[:, 0], out=(b, c))
```

Since `b` and `c` need to be reallocated anyway (to a tensor of 3 elements), isn't it better to do so with the suggested `strides`, restriding these tensors (possibly making the operation faster)?

**Comment by bdhirsh** [comment]
> Just one thing here: at::Tensor::resize_ doesn't take strides (up to this PR). 

Ah yep whoops, I misspoke.

> That would be a solution, but is that the behavior we want? I mean, consider the following example:
> 
> ```python
> x = torch.randn(3, 5, dtype=torch.float)
> b = torch.empty(0)
> c = torch.empty(0, dtype=torch.long)
> torch.sort(x[:, 0], out=(b, c))
> ```
> 
> Since `b` and `c` need to be reallocated anyway (to a tensor of 3 elements), isn't it better to do so with the suggested `strides`, restriding these tensors (possibly making the operation faster)?

Hmm so, what I didn't realize when we initially had to revert the PR - wouldn't that be a change in the behavior of `sort`? As pointed out above, when you run that example `torch.sort()` will set `b` and `c` to be contiguous today.

I think the high bit is probably avoiding a behavior change in the current semantics of sort (maybe I'm missing something though). But that also seems kind of reasonable - do we really want to allocate 5x as much memory as we really need to store the output in the above example, just to maintain strides? Would the remaining memory all just be uninitialized?

**Comment by ngimel** [comment]
With very rare exceptions, we are returning contiguous tensors (and would not allocate 5x memory as in the example above), and we should stick to this behavior. 
For pointwise operations, we are trying to preserve the input permutation, but I'm not sure if we can do it efficiently in `sort` so we can skip this discussion for now, and always return contiguous outputs from sort.

**Comment by ysiraichi** [comment]
> Hmm so, what I didn't realize when we initially had to revert the PR - wouldn't that be a change in the behavior of sort? As pointed out above, when you run that example `torch.sort()` will set `b` and `c` to be contiguous today.

Hmm. Not really. Today (without this PR), `sort` will return contiguous outputs only if the size of the given dimension is less than a threshold (1024 for `long`). Otherwise, it calls `empty_strided` to get strided tensors.

> but I'm not sure if we can do it efficiently in sort so we can skip this discussion for now, and always return contiguous outputs from sort.

Got it. I think this solves the problem with `sort`, then.

But, I still think we should do something about how structured kernels resize/restride the outputs. As is, problems similar to #65863 may pop out somewhere else, again. After discussing with @Lezcano, I think it would probably make sense to always use the given strides (semantically, and for [some linear algebra kernel implementations](https://github.com/pytorch/pytorch/blob/42749fd333438d34f2d4c7f85a6587035e458216/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1625-L1646) that have particular strides requirements). What do you think?

**Comment by Lezcano** [comment]
On the "always use the provided strides", I think that's the right way to go. Otherwise, there's very little point in asking to set the strides to something in the `set_output` method and then having to check that the strides are indeed correct inside the `_out` function. Note that this should apply not only to the functional variant, but also to the `_out` variant, where the inputs are provided by the user and might have incorrect strides. In this case, new inputs with the correct strides should be provided and passed to the `_out` function, and then, after the function is executed, the results should be copied into the tensors originally provided by the user. Note that we are currently implementing this pattern in `linalg.`:
https://github.com/pytorch/pytorch/blob/42749fd333438d34f2d4c7f85a6587035e458216/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1625-L1646

So, to summarise, in my opinion there should be two options:
- The strides for an output are not provided: C-contig is selected by default if a tensor needs to be created, and no guarantees are given on the strides of the given tensors at the start of `TORCH_IMPL_FUNC`.
- A stride is provided: There invariant of the input tensors having the right size AND strides is respected at the start of `TORCH_IMPL_FUNC`.

As a side note, it might be possible to not instantiate new tensors, but simply doing some sort of inplace reshuffling to be able to restride the data. I do not know whether these optimisations can even be done, and we certainly do not implement them when calling external libraries in `linalg`.

**Comment by ngimel** [comment]
>Hmm. Not really. Today (without this PR), sort will return contiguous outputs only if the size of the given dimension is less than a threshold (1024 for long). Otherwise, it calls empty_strided to get strided tensors.
It's true, it calls empty_strided but it always returns memory-dense tensor, empty_strided preserves permutation, but not the actual strides.

**Comment by ysiraichi** [comment]
> It's true, it calls empty_strided but it always returns memory-dense tensor, empty_strided preserves permutation, but not the actual strides.

Ah, you are totally right! Thanks for clarifying that.

So, here's what I'm thinking:
- Make `set_output` enforce the given sizes and strides to the output tensor (as suggested by @Lezcano)
- Always return contiguous outputs for `sort`

@ngimel @bdhirsh @Lezcano 
What do you think?

**Comment by ngimel** [comment]
Can we preserve current sort behavior? With current behavior, outputs are not necessarily contiguous, mimic the input permutation, and avoid extra copies if possible, which seems like desirable state.

**Comment by bdhirsh** [comment]
Summmarizing / confirming my understanding of the discussion above:

(1) the structured kernel port for `at::sort` should preserve the existing striding behavior: outputs are not necessarily contiguous, mimic the input permutation, *but* the output never wastes space - it always takes up the same # bytes as a corresponding contiguous tensor.

(2) Structured kernels should respect strides passed into `set_output()` **except** when it would result in allocating a wasteful amount of storage. This is **already the case** (am I missing something?). I can't think of an easy way to progamatically avoid problems like  #65863, where `at::sort` ends up trying to pass in strides to `set_output()` that would involve wastefully allocated space. I'm also thinking of this problem as independent of @Lezcano's proposal.

(3) @Lezcano's proposal: there are a lot of kernels in pytorch that involve allocating a temporary tensor to perform the computation in (with different strides that make it more efficient), and then copying the result back into the output tensor. That sounds useful + unrelated to #65863

**Comment by Lezcano** [comment]
That summary is perfect!
Just a small nit. We do not need the strides because of speed, but because of correctness, as the external libraries we're calling will perform the computations expecting certain strides.

**Comment by Lezcano** [comment]
On 2, the vector representation of the tensor has holes if after sorting the strides decreasingly and after removing the zero strided dimensions, the vector of strides that you get is not contiguous.

I believe that a solution to this would be to:
- Sort the vector of strides and get the permutation that sorted it.
- Remove the zero-strided dims
- Compute the strides that would make the remaining dimensions contiguous
- Apply the inverse permutation to these strides

I think that these would be the best strides to use in these cases.
Of course, this can be implemented with a number of shortcuts that would avoid doing unnecessary work when possible. For example, if the tensor or its transpose are contiguous or channel contiguous or any of these, there's nothing to do. Same happens if after sorting the strides, it turns out that the sorted strides are contiguous under the definition above. In this case there's no need to compute the inverse permutation.

All this could be encapsulated in a function that given the original tensor, it returns "the best strides" according to the definition above.

I would assume that, provided that TI implements suitable optimisations, this could be the best way to go about things. cc @peterbell10  on this as he might know more about this.

**Comment by peterbell10** [comment]
`TensorIterator` does something very similar:
1. Find dimension permutation that optimizes iteration order (i.e. sorting the strides, but considering all operands)
2. Create strides for a contiguous tensor of the permuted output shape
3. Invert the dimension permutation and pass to `set_output`

I'll add that I don't think `set_output` should replicate this behaviour, otherwise for `TensorIterator` you would be calculating it twice. Instead, I think having `sort` call a helper function `compute_dense_strides_like(shape, strides)` and passing those strides to `set_output` would make the most sense.

**Comment by peterbell10** [comment]
Actually, that helper already exists: `infer_dense_strides`.

**Comment by ysiraichi** [comment]
@peterbell10 @Lezcano @ngimel @bdhirsh 
Thank you for your feedback.

Basing on @bdhirsh summary, here's what I plan to do:
1. make `resize_output` comply to the strides passed to it
2. make `at::sort`call `at::infer_dense_strides` (preserving current behavior)

> Structured kernels should respect strides passed into `set_output()` except when it would result in allocating a wasteful amount of storage

I think it would be better to respect the given strides, instead of calling `at::infer_dense_strides` implicitly inside `resize_output`. The idea being that the function should be called by the kernels that require it (so that we don't get any surprises).

What do you think?

**Comment by Lezcano** [comment]
> I think it would be better to respect the given strides, instead of calling at::infer_dense_strides implicitly inside resize_output

I think so as well.

**Comment by bdhirsh** [comment]
> make `at::sort` call `at::infer_dense_strides` (preserving current behavior)

> I think it would be better to respect the given strides, instead of calling at::infer_dense_strides implicitly inside resize_output. The idea being that that function should be called by the kernels that require it (so that we don't get any surprises).

Totally agreed - what I meant is that I think it's ok to error out in the wasteful storage case (which is what happens today. Maybe that error message could be improved though). I don't think we should transparently call `infer_dense_strides` inside of `set_output()` to "fix" things.

File: tools/codegen/dest/register_dispatch_key.py [hunk 4]
```diff
@@ -73,17 +73,13 @@ def gen_resize_out_helper(backend_index: BackendIndex) -> List[str]:
...
-  if (resized) {
-    if (!strides.empty()) {
-      TORCH_INTERNAL_ASSERT(!options.memory_format_opt().has_value());
-      at::native::as_strided_(out, sizes, strides);
-    } else if (options.memory_format_opt().has_value()) {
-      out.unsafeGetTensorImpl()->empty_tensor_restride(*options.memory_format_opt());
-    }
+  c10::optional<IntArrayRef> expanded_strides = strides.empty() ? c10::nullopt
+      : c10::optional<IntArrayRef>(strides);
+  const bool resized = at::native::resize_output(out, sizes, expanded_strides);
```
Comment (bdhirsh) [inline tools/codegen/dest/register_dispatch_key.py:78]: Is this for @Lezcano 's proposal? If we call something like:
```
a = contiguous_tensor()  # row major strides
torch.linalg_op(..., out=a) # suppose a has the right size, but the linalg_op needs to operate on column major strides
```

Then I don't think we want to actually modify `a`'s strides. out= ops shouldn't modify the metadata on the out= tensor; instead, we need to construct an intermediate tensor with the right strides and pass that into the IMPL func, and then copy back the result to the user's out= tensor.

That's my understanding of @Lezcano's idea. I think it should be done in a separate PR, since it's pretty independent from the `at::sort` issue addressed in this stack (and won't require changes to `resize_()`, but will require changes to the structured kernel codegen). Let me know if you agree!

Comment (Lezcano) [reply tools/codegen/dest/register_dispatch_key.py:78]: That was exactly my idea, yep. I also agree that the strides of the out tensor should not be changed.

Comment (ysiraichi) [reply tools/codegen/dest/register_dispatch_key.py:78]: > Is this for @Lezcano 's proposal?

No, it was for my previous proposal.

> out= ops shouldn't modify the sizes/strides on the out= tensor; 

Ideally, right!? Currently, it does modify the sizes, even though we do throw a warning.

> I think it should be done in a separate PR

That makes sense. So, I'm going to close both PRs in this stack and open a new one for `at::sort`. I'm also opening a new issue just for tracking @Lezcano idea.
